### PR TITLE
Fix invalid menu_item_parent

### PR DESCRIPTION
**Pull request by azaozz**
Trac ticket: https://core.trac.wordpress.org/ticket/57169

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

**Comment by azaozz** [comment]
This should reset the invalid `_menu_item_menu_item_parent` meta whenever the menu is saved. It also contains the runtime fix as mentioned in https://github.com/WordPress/wordpress-develop/pull/3649#issuecomment-1322728338.

Ideally this should go in after https://github.com/WordPress/wordpress-develop/pull/3649 (https://core.trac.wordpress.org/ticket/56926).

File: src/wp-includes/nav-menu.php [hunk 1]
```diff
@@ -1273,3 +1278,29 @@ function wp_map_nav_menu_locations( $new_nav_menu_locations, $old_nav_menu_locat
 
 	return $new_nav_menu_locations;
 }
+
+/**
+ * Private function to reset menu_item_parent when it is the same as the menu item ID.
+ * For use before saving `_menu_item_menu_item_parent` in nav-menus.php.
```
Comment (peterwilsoncc) [inline src/wp-includes/nav-menu.php]:
```suggestion
 * Prevents menu items from being their own parent.
 *
 * Resets menu_item_parent to 0 when the parent is set to the item itself.
 * For use before saving `_menu_item_menu_item_parent` in nav-menus.php.
```

**Review by peterwilsoncc — APPROVED** [review]

File: src/wp-includes/nav-menu-template.php [hunk 2]
```diff
@@ -199,7 +199,8 @@ function wp_nav_menu( $args = array() ) {
 	$menu_items_with_children = array();
 	foreach ( (array) $menu_items as $menu_item ) {
 		// Fix invalid `menu_item_parent`. See: https://core.trac.wordpress.org/ticket/56926.
-		if ( (int) $menu_item->ID === (int) $menu_item->menu_item_parent ) {
+		// Compare as strings. Plugins may change the ID to string.
+		if ( "$menu_item->ID" === "$menu_item->menu_item_parent" ) {
```
Comment (peterwilsoncc) [inline src/wp-includes/nav-menu-template.php]:
```suggestion
		if ( (string) $menu_item->ID === (string) $menu_item->menu_item_parent ) {
```

Opinionated: this seems clearer to me than quotes -- the effect is the same.

Comment (azaozz) [reply src/wp-includes/nav-menu-template.php]: > this seems clearer to me than quotes

Yep, it is. Casting to string was the first thing I did too :) 

However it changes/casts both `$menu_item->ID` and `$menu_item->menu_item_parent` to strings which might perhaps introduce a bug somewhere else? Both of these are usually numeric (post IDs for the current post and parent post). 

Another possible thing would be to make new vars for these, although seems perhaps too wordy:
```
$menu_item_id     = strval( $menu_item->ID );
$menu_item_parent = strval( $menu_item->menu_item_parent );

if ( $menu_item_id === $menu_item_parent ) {
.....
```

Yet another way to write this would be:
```
if ( strval( $menu_item->ID ) === strval( $menu_item->menu_item_parent ) ) {
.....
```

Comment (peterwilsoncc) [reply src/wp-includes/nav-menu-template.php]: > However it changes/casts both `$menu_item->ID` and `$menu_item->menu_item_parent` to strings which might perhaps introduce a bug somewhere else? Both of these are usually numeric (post IDs for the current post and parent post).

Hmm, I wasn't seeing that in my reduced case. Probably because I missed something by reducing it :) 

Let's keep it as is and include an explanation in the comment.

```php
/*
 * Compare as strings as plugins may change the ID to string.
 * To avoid modifying the object, these items are quoted rather than cast to (string).
 */
```

Or something like that anyway.

Comment (azaozz) [reply src/wp-includes/nav-menu-template.php]: > include an explanation in the comment.

Yes, sounds good.

Also... Re-reading this today thinking this is the most readable and self-documenting code:
```if ( strval( $menu_item->ID ) === strval( $menu_item->menu_item_parent ) )```

`strval()` is equivalent to using double-quotes, and is used at a few other places in core.

**Review by peterwilsoncc — APPROVED** [review]
Thanks Ozz, LGTM
